### ISSUE TITLE
Clarify dependencies

### ISSUE THREAD

**Issue by eddyashton** [microsoft/CCF#3517]
The build-time dependency graph between components in CCF has grown messier over time. We originally had `README.md`s in some `src/` folders indicating their intended dependencies, but these have rotted over time.

We should re-evaluate what dependencies we have and can enforce, to keep the code base navigable.

A few high-level goals to start with:
- Public headers should not include any private headers (#2428).
- 3rd party/STL libs are generally useable anywhere/visible everywhere (for instance `nlohmann/json` is used throughout the codebase, including in core containers), with a few exceptions: We should limit visibility of anything platform-specific (including SGX and host-only libraries).
- `ds/` should not depend on anything else. This is a place for generic containers and shims that may be used by multiple other components.
- We should aim to have a set of pure components which depend only on `ds/`. For instance `crypto/`, `tls/`, and `http/` could be written like this - generic protocol implementations and helpers, stripped of platform-integration features.
- Where we have tightly-coupled systems like `kv/` and `consensus/`, we should try to document this coupling to ensure it is required.

**Comment by eddyashton** [comment]
The repo currently contains 3 READMEs which claim the following dependencies:

```
ds/README.md
# Data Structures

This directory contains data structures and utilities used by the rest of the code. No files
in this directory should include files from any other directory.

kv/README.md
# Transactional Key Value Store

This directory implements a transactional key value store. The only directory
it should depend on is `ds`.

crypto/README.md
# Crypto

This directory implements several cryptography helper types. The only directory
it should depend on is `ds`.
```

None of these dependencies are currently enforced. The actual dependencies between the folders under `src/` look like this:
![image](https://user-images.githubusercontent.com/6000239/153004744-5694a0d0-4c31-4dd5-a791-d97d47251a64.png)

Dotted lines show mutual/circular dependencies. Initial thoughts:
- We're aware of the mutual dependency between `kv/` and `consensus/`, and believe it's extremely hard to break
- We should try to break the dependency from `ds/` to `crypto/`
- A surprising number of things rely directly on `crypto/`, I wonder if it's worth splitting this
- This shows we do already have a few 'clean' components at the _bottom_ of the tree, which are only consuming other libraries (`clients/`, `host/`, `apps/`), we should enforce that distinction while we have it

**Comment by achamayou** [comment]
Updated from `main` at [`488f8da`](https://github.com/microsoft/CCF/commit/488f8da0e339a4ac1752b0f27a3b34792ec47ed4) (2026-07-10).

This is the same source-folder view as the original diagram: each node is a top-level directory under `src/` containing C/C++ source or headers, and `A -> B` means that a non-test source/header in `src/A` directly includes a header owned by `B` (through `src/B`, `include/ccf/B`, or a relative path). Tests, benchmarks, fuzzers, and third-party/STL dependencies are omitted. Dashed arrows participate in a dependency cycle; solid arrows do not. The subgraph groupings are only for readability, not dependency layers.

```mermaid
flowchart TB
  subgraph consumers["Acyclic consumers"]
    direction LR
    apps["apps/"]
    clients["clients/"]
  end

  subgraph runtime["Node and execution"]
    direction LR
    host["host/"]
    enclave["enclave/"]
    node["node/"]
    endpoints["endpoints/"]
    js["js/"]
    indexing["indexing/"]
    snapshots["snapshots/"]
  end

  subgraph state["Replicated state and service"]
    direction LR
    consensus["consensus/"]
    kv["kv/"]
    service["service/"]
  end

  subgraph protocols["Protocols and transport"]
    direction LR
    http["http/"]
    tls["tls/"]
    quic["quic/"]
    tcp["tcp/"]
    udp["udp/"]
  end

  subgraph support["Foundation and support"]
    direction LR
    common["common/"]
    crypto["crypto/"]
    ds["ds/"]
    pal["pal/"]
    tasks["tasks/"]
    cose["cose/"]
  end

  apps --> js
  clients --> crypto
  clients --> ds
  clients --> http
  clients --> tls
  common -.-> consensus
  common -.-> crypto
  common -.-> ds
  common -.-> node
  common -.-> pal
  common -.-> service
  consensus -.-> crypto
  consensus -.-> ds
  consensus -.-> enclave
  consensus -.-> kv
  consensus -.-> node
  consensus -.-> pal
  consensus -.-> service
  crypto --> cose
  crypto -.-> ds
  ds -.-> crypto
  ds -.-> pal
  ds -.-> service
  enclave -.-> common
  enclave -.-> crypto
  enclave -.-> ds
  enclave -.-> host
  enclave -.-> http
  enclave -.-> indexing
  enclave -.-> js
  enclave -.-> kv
  enclave -.-> node
  enclave -.-> pal
  enclave -.-> quic
  enclave -.-> service
  enclave -.-> tasks
  enclave -.-> tcp
  enclave -.-> tls
  enclave -.-> udp
  endpoints -.-> crypto
  endpoints -.-> ds
  endpoints -.-> http
  endpoints -.-> kv
  endpoints -.-> node
  endpoints -.-> pal
  endpoints -.-> service
  host -.-> common
  host -.-> consensus
  host -.-> crypto
  host -.-> ds
  host -.-> enclave
  host -.-> http
  host -.-> indexing
  host -.-> kv
  host -.-> node
  host -.-> pal
  host -.-> service
  host -.-> snapshots
  host -.-> tcp
  host -.-> udp
  http -.-> crypto
  http -.-> ds
  http -.-> enclave
  http -.-> host
  http -.-> node
  indexing -.-> crypto
  indexing -.-> ds
  indexing -.-> kv
  indexing -.-> node
  indexing -.-> pal
  js -.-> crypto
  js -.-> ds
  js -.-> endpoints
  js -.-> kv
  js -.-> node
  js -.-> pal
  js -.-> service
  js -.-> tls
  kv -.-> crypto
  kv -.-> ds
  kv -.-> node
  kv -.-> pal
  kv -.-> service
  node -.-> common
  node -.-> consensus
  node --> cose
  node -.-> crypto
  node -.-> ds
  node -.-> enclave
  node -.-> endpoints
  node -.-> host
  node -.-> http
  node -.-> indexing
  node -.-> js
  node -.-> kv
  node -.-> pal
  node -.-> service
  node -.-> snapshots
  node -.-> tasks
  pal -.-> crypto
  pal -.-> ds
  quic -.-> ds
  quic -.-> enclave
  quic -.-> udp
  service -.-> consensus
  service --> cose
  service -.-> crypto
  service -.-> ds
  service -.-> kv
  service -.-> node
  service -.-> pal
  snapshots -.-> consensus
  snapshots -.-> ds
  snapshots -.-> host
  snapshots -.-> http
  tasks -.-> ds
  tcp -.-> ds
  tls -.-> crypto
  tls -.-> ds
  udp -.-> ds
```

At this revision there are 23 components and 120 direct internal dependency edges. `apps/` and `clients/` are the only acyclic consumers, `cose/` is an acyclic sink, and the other 20 components form one strongly connected component. In particular, the original intended boundaries still do not hold: `ds/` directly depends on `crypto/`, `pal/`, and `service/`; `kv/` directly depends on `crypto/`, `ds/`, `node/`, `pal/`, and `service/`; and `crypto/` directly depends on `ds/` and `cose/`. Unlike the 2022 graph, `host/` is no longer a clean outer consumer; it is now part of the cyclic core.

**Comment by achamayou** [comment]
Rerun against `main` at [`9751a52`](https://github.com/microsoft/CCF/commit/9751a52593614f7f64a51510dc9606ec2458bfe9) (2026-09-02), using the same include-level methodology and exclusions as the July analysis.

There are now **24 components and 116 direct edges**. Crucially, the former 20-component cycle has split into:

- A 13-component cycle: `common`, `consensus`, `enclave`, `endpoints`, `host`, `http`, `indexing`, `js`, `kv`, `node`, `quic`, `service`, `snapshots`.
- A 3-component cycle: `crypto`, `ds`, `pal`.
- Eight acyclic components: `apps`, `cose`, `msgpack`, `tasks`, `tcp`, `threading`, `tls`, `udp`.

```mermaid
flowchart TB
  subgraph consumers["Acyclic consumers"]
    apps["apps/"]
  end

  subgraph runtime["Node and execution"]
    host["host/"]
    enclave["enclave/"]
    node["node/"]
    endpoints["endpoints/"]
    js["js/"]
    indexing["indexing/"]
    snapshots["snapshots/"]
  end

  subgraph state["Replicated state and service"]
    consensus["consensus/"]
    kv["kv/"]
    service["service/"]
  end

  subgraph protocols["Protocols and transport"]
    http["http/"]
    tls["tls/"]
    quic["quic/"]
    tcp["tcp/"]
    udp["udp/"]
  end

  subgraph support["Foundation and support"]
    common["common/"]
    crypto["crypto/"]
    ds["ds/"]
    pal["pal/"]
    tasks["tasks/"]
    cose["cose/"]
    msgpack["msgpack/"]
    threading["threading/"]
  end

  apps --> js
  common -.-> consensus
  common --> crypto
  common --> ds
  common -.-> node
  common --> pal
  common -.-> service
  consensus --> crypto
  consensus --> ds
  consensus -.-> enclave
  consensus -.-> kv
  consensus -.-> node
  consensus --> pal
  consensus -.-> service
  crypto --> cose
  crypto -.-> ds
  ds -.-> pal
  enclave -.-> common
  enclave --> crypto
  enclave --> ds
  enclave -.-> host
  enclave -.-> http
  enclave -.-> indexing
  enclave -.-> js
  enclave -.-> kv
  enclave -.-> node
  enclave --> pal
  enclave -.-> quic
  enclave -.-> service
  enclave --> tasks
  enclave --> tcp
  enclave --> tls
  enclave --> udp
  endpoints --> crypto
  endpoints --> ds
  endpoints -.-> http
  endpoints -.-> kv
  endpoints -.-> node
  endpoints --> pal
  endpoints -.-> service
  host -.-> common
  host -.-> consensus
  host --> crypto
  host --> ds
  host -.-> enclave
  host -.-> http
  host -.-> indexing
  host -.-> kv
  host -.-> node
  host --> pal
  host -.-> service
  host -.-> snapshots
  host --> tcp
  host --> udp
  http --> crypto
  http --> ds
  http -.-> enclave
  http -.-> host
  http -.-> node
  http --> pal
  indexing --> crypto
  indexing --> ds
  indexing -.-> kv
  indexing -.-> node
  indexing --> pal
  js --> crypto
  js --> ds
  js -.-> endpoints
  js -.-> kv
  js -.-> node
  js --> pal
  js -.-> service
  js --> tls
  kv --> crypto
  kv --> ds
  kv -.-> node
  kv --> pal
  kv -.-> service
  node -.-> common
  node -.-> consensus
  node --> cose
  node --> crypto
  node --> ds
  node -.-> enclave
  node -.-> endpoints
  node -.-> host
  node -.-> http
  node -.-> indexing
  node -.-> js
  node -.-> kv
  node --> pal
  node -.-> service
  node -.-> snapshots
  node --> tasks
  pal -.-> crypto
  pal -.-> ds
  quic --> ds
  quic -.-> enclave
  quic --> udp
  service -.-> consensus
  service --> cose
  service --> crypto
  service --> ds
  service -.-> kv
  service -.-> node
  service --> pal
  snapshots -.-> consensus
  snapshots --> ds
  snapshots -.-> host
  snapshots -.-> http
  tasks --> ds
  tasks --> pal
  tcp --> ds
  tls --> crypto
  tls --> ds
  udp --> ds
```

Dashed arrows participate in a cycle.

Compared with the July baseline at `488f8da`:

- `clients/` was removed; `msgpack/` and `threading/` were added as internally independent components.
- Moving `cli_helper.h` from `ds/` to `common/` removed `ds -> crypto` and `ds -> service`, splitting the giant cycle.
- `http -> pal` and `tasks -> pal` were added.
- `ds/` now depends only on `pal/`, but remains cyclic through `ds -> pal -> crypto -> ds`.
- `kv/` remains coupled to `crypto`, `ds`, `node`, `pal`, and `service`; `host/` remains in the main cyclic core.

**Comment by achamayou** [comment]
Rerun against `main` at [`09568fb`](https://github.com/microsoft/CCF/commit/09568fbb623eca962c0887c0aa84be4f8b0ebd75) (2026-09-04), using the deterministic include analysis now implemented by `scripts/check-source-dependencies.py` and the same production-source exclusions as the previous updates.

There are now **23 components and 104 direct internal edges**. The remaining cyclic core is a single 12-component SCC:

- `common`, `consensus`, `enclave`, `endpoints`, `host`, `http`, `indexing`, `js`, `kv`, `node`, `service`, `snapshots`.

The other 11 components are acyclic: `apps`, `cose`, `crypto`, `ds`, `msgpack`, `pal`, `tasks`, `tcp`, `threading`, `tls`, and `udp`.

Compared with the 2026-09-02 baseline at `9751a52`:

- #8265 removed eight edges into `pal`, including the last outgoing edge from `ds`. This eliminated the separate `crypto -> ds -> pal -> crypto` cycle. The resulting enforced foundation is `ds -> {}`, `crypto -> {cose, ds}`, and `pal -> {crypto, ds}`.
- #8265 also added the dependency checker to `scripts/includes-checks.sh`. It currently enforces declared dependencies for `crypto`, `ds`, and `pal`; it does not yet prove that the full repository is a DAG.
- #8271 removed the experimental `quic` component and its four graph edges, reducing the main SCC from 13 components to 12.

Overall, the graph moved from 24 components / 116 edges / 16 cyclic components to 23 components / 104 edges / 12 cyclic components.

Dashed arrows below participate in the remaining cycle; solid arrows do not.

```mermaid
flowchart TB
  subgraph consumers["Acyclic consumers"]
    apps["apps/"]
  end

  subgraph runtime["Node and execution"]
    host["host/"]
    enclave["enclave/"]
    node["node/"]
    endpoints["endpoints/"]
    js["js/"]
    indexing["indexing/"]
    snapshots["snapshots/"]
  end

  subgraph state["Replicated state and service"]
    consensus["consensus/"]
    kv["kv/"]
    service["service/"]
  end

  subgraph protocols["Protocols and transport"]
    http["http/"]
    tls["tls/"]
    tcp["tcp/"]
    udp["udp/"]
  end

  subgraph support["Foundation and support"]
    common["common/"]
    crypto["crypto/"]
    ds["ds/"]
    pal["pal/"]
    tasks["tasks/"]
    cose["cose/"]
    msgpack["msgpack/"]
    threading["threading/"]
  end

  apps --> js
  common -.-> consensus
  common --> crypto
  common --> ds
  common -.-> node
  common --> pal
  common -.-> service
  consensus --> crypto
  consensus --> ds
  consensus -.-> enclave
  consensus -.-> kv
  consensus -.-> node
  consensus -.-> service
  crypto --> cose
  crypto --> ds
  enclave -.-> common
  enclave --> crypto
  enclave --> ds
  enclave -.-> host
  enclave -.-> http
  enclave -.-> indexing
  enclave -.-> js
  enclave -.-> kv
  enclave -.-> node
  enclave -.-> service
  enclave --> tasks
  enclave --> tcp
  enclave --> tls
  enclave --> udp
  endpoints --> crypto
  endpoints --> ds
  endpoints -.-> http
  endpoints -.-> kv
  endpoints -.-> node
  endpoints -.-> service
  host -.-> common
  host -.-> consensus
  host --> crypto
  host --> ds
  host -.-> enclave
  host -.-> http
  host -.-> indexing
  host -.-> kv
  host -.-> node
  host --> pal
  host -.-> service
  host -.-> snapshots
  host --> tcp
  host --> udp
  http --> crypto
  http --> ds
  http -.-> enclave
  http -.-> host
  http -.-> node
  indexing --> crypto
  indexing --> ds
  indexing -.-> kv
  indexing -.-> node
  js --> crypto
  js --> ds
  js -.-> endpoints
  js -.-> kv
  js -.-> node
  js --> pal
  js -.-> service
  js --> tls
  kv --> crypto
  kv --> ds
  kv -.-> node
  kv -.-> service
  node -.-> common
  node -.-> consensus
  node --> cose
  node --> crypto
  node --> ds
  node -.-> enclave
  node -.-> endpoints
  node -.-> host
  node -.-> http
  node -.-> indexing
  node -.-> js
  node -.-> kv
  node --> pal
  node -.-> service
  node -.-> snapshots
  node --> tasks
  pal --> crypto
  pal --> ds
  service -.-> consensus
  service --> cose
  service --> crypto
  service --> ds
  service -.-> kv
  service -.-> node
  service --> pal
  snapshots -.-> consensus
  snapshots --> ds
  snapshots -.-> host
  snapshots -.-> http
  tasks --> ds
  tcp --> ds
  tls --> crypto
  tls --> ds
  udp --> ds
```

**Comment by achamayou** [comment]
## 2026-09-09: HTTP is acyclic; snapshots should be next

Current `main` is [`ab11900`](https://github.com/microsoft/CCF/commit/ab119009530e6fa09f8dc781624914a42ee475f0). The source dependency inputs are unchanged from the locally analysed [`7b31f9e`](https://github.com/microsoft/CCF/commit/7b31f9e2cb266ba09c9ecfb07a6faba84733a11e): the intervening #8277 adds the Lean model and associated metadata, not C/C++ dependencies.

### Current state

| Revision / proposed step | Source components | Direct source-component edges | Components in cycles | Acyclic components | Components with an outgoing dependency policy |
| --- | ---: | ---: | ---: | ---: | ---: |
| 2026-09-04, `09568fb` | 23 | 104 | 12 | 11 | 3 |
| 2026-09-09, `ab11900` | 25 | 106 | 11 | 14 | 14 |
| Proposed snapshots cut, against this baseline | 25 | 104 | 10 | 15 | 15 |

There is now **one 11-component strongly connected component (SCC)**:

`common`, `consensus`, `enclave`, `endpoints`, `host`, `indexing`, `js`, `kv`, `node`, `service`, `snapshots`.

The **14 acyclic components** are:

`apps`, `cose`, `crypto`, `ds`, `http`, `http_client`, `msgpack`, `pal`, `tasks`, `tcp`, `threading`, `tls`, `udp`, `uv`.

All 14 currently acyclic components now have explicit outgoing dependency policies. This is progress toward a component DAG, **not a claim that the repository is already a DAG**.

### Progress since the last update

- #8278 added policies for the previously unguarded acyclic components.
- #8281 extracted the generic libuv ownership wrapper into `uv`, removing `http -> host`.
- #8284 moved HTTP runtime/session adapters into `enclave`, removing `http -> enclave` and `http -> node`. Together with #8281, this took `http` out of the cyclic core.
- #8285 extracted the libcurl/libuv client into `http_client`. Its callers now depend on that component rather than treating it as part of the HTTP protocol implementation.

The current policies include:

```text
ds          -> {}
crypto      -> {cose, ds}
pal         -> {crypto, ds}
uv          -> {}
http        -> {ccf-api, crypto, ds}
http_client -> {ccf-api, ds, uv}
```

Compared with September 4, the four removed source-component edges are `host -> http` and `http -> {enclave, host, node}`. The six added edges are `host -> {http_client, uv}`, `http_client -> {ds, uv}`, `node -> http_client`, and `snapshots -> http_client`. The net increase of two edges reflects the new reusable components, not a new cycle.

### Next move: make `snapshots` acyclic in one small PR

Its only outgoing edges into the cyclic core are:

| Edge to remove | All direct include witnesses | Actual dependency |
| --- | --- | --- |
| `snapshots -> host` | [`filenames.h:6`](https://github.com/microsoft/CCF/blob/ab119009530e6fa09f8dc781624914a42ee475f0/src/snapshots/filenames.h#L6), [`snapshot_writer.h:8`](https://github.com/microsoft/CCF/blob/ab119009530e6fa09f8dc781624914a42ee475f0/src/snapshots/snapshot_writer.h#L8) | The generic `TimeBoundLogger` helper |
| `snapshots -> consensus` | [`snapshot_writer.h:6`](https://github.com/microsoft/CCF/blob/ab119009530e6fa09f8dc781624914a42ee475f0/src/snapshots/snapshot_writer.h#L6) | `consensus::Index` in the two `persist_snapshot` index parameters; no consensus operations |

**Do both cuts together.** Removing either edge alone leaves all 11 components cyclic. No single-edge deletion in the current graph reduces the number of cyclic components. Among the two-edge cuts that do, this pair has the fewest direct include witnesses: three. That is an ownership signal, not an estimate that the complete patch is only three lines.

1. **Move `TimeBoundLogger` from `src/host/time_bound_logger.h` to `src/ds/time_bound_logger.h`.** Give the generic helper an appropriate `ccf::ds` namespace and update all host and snapshot consumers. Its dependencies are already just `ds`, the standard library, and `fmt`. Preserve the single shared `default_max_time`, including the `slow_io_logging_threshold` assignment in [`host/run.cpp`](https://github.com/microsoft/CCF/blob/ab119009530e6fa09f8dc781624914a42ee475f0/src/host/run.cpp#L1080-L1081), and preserve formatting and logging thresholds. Do not duplicate the helper or introduce a `ds -> host` forwarding include.
2. **Replace the writer's `consensus::Index` parameters with the existing `ccf::SeqNo` from `ccf/tx_id.h`.** Both are exactly `uint64_t` today: [`Index`](https://github.com/microsoft/CCF/blob/ab119009530e6fa09f8dc781624914a42ee475f0/src/consensus/ledger_enclave_types.h#L13), [`SeqNo`](https://github.com/microsoft/CCF/blob/ab119009530e6fa09f8dc781624914a42ee475f0/include/ccf/tx_id.h#L36). Remove the ledger ringbuffer header from `snapshot_writer.h`; do not move consensus machinery into `ds`, create another sequence-number type, or change snapshot bytes/filenames.
3. **Enforce the resulting boundary in `scripts/source-dependencies.json`:** `"snapshots": ["ccf-api", "ds", "http", "http_client"]`. Keep the existing foundation policies unchanged. Resolve any missing direct includes exposed by the moves, rather than relying on unrelated headers to make them compile.
4. **Keep the PR structural and demonstrate the intended boundary.** Add standalone-header compilation coverage using the existing unit-target pattern, plus focused coverage for the relocated timing helper. Build the host/runtime consumers and the affected `logger_test`, `ledger_test`, `files_cleanup_test`, `indexing_test`, and `snapshotter_test` targets. Run those unit tests and the existing `recovery_test` e2e path. Preserve the writer's exclusive creation, complete writes, fsync/rename sequence, and failure handling, and retain the snapshotter's commit/rollback behaviour. Run the include/dependency and applicable formatting/ASCII checks before pushing.

**Acceptance criteria:** no production include from `snapshots` into `host` or `consensus`; `snapshots` has only the declared outgoing dependencies; the foundation policies still hold; and the recomputed source graph has **15 acyclic components and one 10-component SCC**. At this baseline the source edge count should fall from **106 to 104**. These are projections for the proposed change, not landed results.

This is preferable to another isolated one-include cleanup such as moving `HistoricalTransactionFetcher`: removing `indexing -> node` alone still leaves `indexing -> kv -> node -> indexing`, so it does not shrink the SCC. It also avoids bundling the broader configuration, RPC, and KV/consensus work already in flight in #8309, #8117, and #8190.

<details>
<summary>Full refreshed source-component graph (25 components, 106 edges)</summary>

As in the previous updates, `A -> B` means a production source/header in `src/A` directly includes a header owned by `B`. Dashed arrows participate in the remaining SCC; solid arrows do not. Groupings are only for readability.

```mermaid
flowchart TB
  subgraph consumers["Acyclic consumers"]
    apps["apps/"]
  end

  subgraph runtime["Node and execution"]
    host["host/"]
    enclave["enclave/"]
    node["node/"]
    endpoints["endpoints/"]
    js["js/"]
    indexing["indexing/"]
    snapshots["snapshots/"]
  end

  subgraph state["Replicated state and service"]
    consensus["consensus/"]
    kv["kv/"]
    service["service/"]
  end

  subgraph protocols["Protocols and transport"]
    http["http/"]
    http_client["http_client/"]
    tls["tls/"]
    tcp["tcp/"]
    udp["udp/"]
  end

  subgraph support["Foundation and support"]
    common["common/"]
    crypto["crypto/"]
    ds["ds/"]
    pal["pal/"]
    tasks["tasks/"]
    cose["cose/"]
    msgpack["msgpack/"]
    threading["threading/"]
    uv["uv/"]
  end

  apps --> js
  common -.-> consensus
  common --> crypto
  common --> ds
  common -.-> node
  common --> pal
  common -.-> service
  consensus --> crypto
  consensus --> ds
  consensus -.-> enclave
  consensus -.-> kv
  consensus -.-> node
  consensus -.-> service
  crypto --> cose
  crypto --> ds
  enclave -.-> common
  enclave --> crypto
  enclave --> ds
  enclave -.-> host
  enclave --> http
  enclave -.-> indexing
  enclave -.-> js
  enclave -.-> kv
  enclave -.-> node
  enclave -.-> service
  enclave --> tasks
  enclave --> tcp
  enclave --> tls
  enclave --> udp
  endpoints --> crypto
  endpoints --> ds
  endpoints --> http
  endpoints -.-> kv
  endpoints -.-> node
  endpoints -.-> service
  host -.-> common
  host -.-> consensus
  host --> crypto
  host --> ds
  host -.-> enclave
  host --> http_client
  host -.-> indexing
  host -.-> kv
  host -.-> node
  host --> pal
  host -.-> service
  host -.-> snapshots
  host --> tcp
  host --> udp
  host --> uv
  http --> crypto
  http --> ds
  http_client --> ds
  http_client --> uv
  indexing --> crypto
  indexing --> ds
  indexing -.-> kv
  indexing -.-> node
  js --> crypto
  js --> ds
  js -.-> endpoints
  js -.-> kv
  js -.-> node
  js --> pal
  js -.-> service
  js --> tls
  kv --> crypto
  kv --> ds
  kv -.-> node
  kv -.-> service
  node -.-> common
  node -.-> consensus
  node --> cose
  node --> crypto
  node --> ds
  node -.-> enclave
  node -.-> endpoints
  node -.-> host
  node --> http
  node --> http_client
  node -.-> indexing
  node -.-> js
  node -.-> kv
  node --> pal
  node -.-> service
  node -.-> snapshots
  node --> tasks
  pal --> crypto
  pal --> ds
  service -.-> consensus
  service --> cose
  service --> crypto
  service --> ds
  service -.-> kv
  service -.-> node
  service --> pal
  snapshots -.-> consensus
  snapshots --> ds
  snapshots -.-> host
  snapshots --> http
  snapshots --> http_client
  tasks --> ds
  tcp --> ds
  tls --> crypto
  tls --> ds
  udp --> ds
```

</details>

### Measurement scope

This uses the scanner functions in [`scripts/check-source-dependencies.py`](https://github.com/microsoft/CCF/blob/ab119009530e6fa09f8dc781624914a42ee475f0/scripts/check-source-dependencies.py) and the exclusions/generated-header list in [`scripts/source-dependencies.json`](https://github.com/microsoft/CCF/blob/ab119009530e6fa09f8dc781624914a42ee475f0/scripts/source-dependencies.json), followed by SCC analysis. It retains the previous updates' source-folder projection: nodes are top-level `src` components containing production C/C++ sources/headers; tests, benchmarks, and fuzz sources are excluded; third-party/STL dependencies and public-only component targets are omitted from the displayed graph.

Public headers under `include/ccf` are resolved as include targets, but their own contents are not separately scanned as source vertices. In particular, `ccf-api` appears in policies but is not a node in this source-folder projection; the other public-only target encountered here is `research`. This is not a linker/build-target or complete transitive public-header graph.

The existing checker enforces configured outgoing dependency allowlists through `scripts/includes-checks.sh`; it does not currently emit this graph or prove global acyclicity. The full DAG goal therefore remains open.

**Comment by achamayou** [comment]
## 2026-09-13: projected after #8338, left to right

**Assuming #8338 merges**, the graph below is measured on GitHub's synthetic merge [`5e32dfb`](https://github.com/microsoft/CCF/commit/5e32dfbc4778018f5c2e377abdb125bc9186b808): `main` at [`31f5f9f`](https://github.com/microsoft/CCF/commit/31f5f9fe14972cda34f7f44e258557f4dc574b51) plus PR head [`cb954b1`](https://github.com/microsoft/CCF/commit/cb954b118c476170f285ac5aaa290027c48bfdb4). This is a projection, not a claim that the PR has landed.

| State | Components | Direct edges | Components in cycles | Acyclic components | Outgoing policies |
| --- | ---: | ---: | ---: | ---: | ---: |
| Current `main` | 25 | 106 | 11 | 14 | 14 |
| After #8338 | 25 | 104 | 10 | 15 | 15 |

The only removed source-component edges are `snapshots -> host` and `snapshots -> consensus`; none are added. `snapshots` leaves the cyclic core, with the enforced boundary `snapshots -> {ccf-api, ds, http, http_client}`. All 15 acyclic components have outgoing dependency policies.

**Legend:** arrows run from an includer to its dependency. Dashed arrows participate in the remaining cycle; solid arrows do not. **Red nodes and red arrows mark the next candidate ownership boundaries to address**, not additional changes assumed in #8338. Green marks `snapshots`, acyclic under the merge assumption. Groupings are for readability, not dependency layers.

```mermaid
flowchart LR
  apps["apps/"]
  snapshots["snapshots/ (post-8338)"]
  subgraph core["Remaining cyclic core (10)"]
    direction LR
    common["common/"]
    consensus["consensus/"]
    enclave["enclave/"]
    endpoints["endpoints/"]
    host["host/"]
    indexing["indexing/"]
    js["js/"]
    kv["kv/"]
    node["node/"]
    service["service/"]
  end
  subgraph protocols["Acyclic protocols and transport"]
    direction LR
    http["http/"]
    http_client["http_client/"]
    tls["tls/"]
    tcp["tcp/"]
    udp["udp/"]
  end
  subgraph support["Acyclic foundation and support"]
    direction LR
    crypto["crypto/"]
    ds["ds/"]
    pal["pal/"]
    tasks["tasks/"]
    cose["cose/"]
    msgpack["msgpack/"]
    threading["threading/"]
    uv["uv/"]
  end

  apps --> js
  common -.-> consensus
  common --> crypto
  common --> ds
  common -.-> node
  common --> pal
  common -.-> service
  consensus --> crypto
  consensus --> ds
  consensus -.-> enclave
  consensus -.-> kv
  consensus -.-> node
  consensus -.-> service
  crypto --> cose
  crypto --> ds
  enclave -.-> common
  enclave --> crypto
  enclave --> ds
  enclave -.-> host
  enclave --> http
  enclave -.-> indexing
  enclave -.-> js
  enclave -.-> kv
  enclave -.-> node
  enclave -.-> service
  enclave --> tasks
  enclave --> tcp
  enclave --> tls
  enclave --> udp
  endpoints --> crypto
  endpoints --> ds
  endpoints --> http
  endpoints -.-> kv
  endpoints -.-> node
  endpoints -.-> service
  host -.-> common
  host -.-> consensus
  host --> crypto
  host --> ds
  host -.-> enclave
  host --> http_client
  host -.-> indexing
  host -.-> kv
  host -.-> node
  host --> pal
  host -.-> service
  host --> snapshots
  host --> tcp
  host --> udp
  host --> uv
  http --> crypto
  http --> ds
  http_client --> ds
  http_client --> uv
  indexing --> crypto
  indexing --> ds
  indexing -.-> kv
  indexing -.-> node
  js --> crypto
  js --> ds
  js -.-> endpoints
  js -.-> kv
  js -.-> node
  js --> pal
  js -.-> service
  js --> tls
  kv --> crypto
  kv --> ds
  kv -.-> node
  kv -.-> service
  node -.-> common
  node -.-> consensus
  node --> cose
  node --> crypto
  node --> ds
  node -.-> enclave
  node -.-> endpoints
  node -.-> host
  node --> http
  node --> http_client
  node -.-> indexing
  node -.-> js
  node -.-> kv
  node --> pal
  node -.-> service
  node --> snapshots
  node --> tasks
  pal --> crypto
  pal --> ds
  service -.-> consensus
  service --> cose
  service --> crypto
  service --> ds
  service -.-> kv
  service -.-> node
  service --> pal
  snapshots --> ds
  snapshots --> http
  snapshots --> http_client
  tasks --> ds
  tcp --> ds
  tls --> crypto
  tls --> ds
  udp --> ds

  classDef opportunity fill:#ffebe9,stroke:#cf222e,stroke-width:3px,color:#82071e
  classDef extracted fill:#dafbe1,stroke:#1a7f37,stroke-width:2px,color:#116329
  class host,indexing opportunity
  class snapshots extracted
  linkStyle 18,56,57,77 stroke:#cf222e,stroke-width:3px
```

### Next targets in red

| Priority | Target | Cyclic edges to address together | Direct include witnesses |
| --- | --- | --- | ---: |
| 1 | `host`: restore an outer consumer | `enclave -> host`, `node -> host` | 5 |
| 2 | `indexing`: separate runtime adapters and the KV implementation | `indexing -> node`, `indexing -> kv` | 5 |

**Start with `host`.** Its remaining incoming dependencies from the cyclic core are all ledger-related: three includes of `host/ledger.h` in the enclave entry/construction path, plus [`ReadLedgerSubsystem`](https://github.com/microsoft/CCF/blob/31f5f9fe14972cda34f7f44e258557f4dc574b51/src/node/rpc/ledger_subsystem.h#L5-L27) and [recovery's ledger filename helpers](https://github.com/microsoft/CCF/blob/31f5f9fe14972cda34f7f44e258557f4dc574b51/src/node/recovery_snapshot_ledger.h#L5-L11) in `node`. Inject the existing [`AbstractReadLedgerSubsystemInterface`](https://github.com/microsoft/CCF/blob/31f5f9fe14972cda34f7f44e258557f4dc574b51/src/node/rpc/ledger_interface.h#L10-L26) instead of the [concrete ledger at the enclave entry point](https://github.com/microsoft/CCF/blob/31f5f9fe14972cda34f7f44e258557f4dc574b51/src/enclave/entry_points.h#L5-L25); keep the concrete adapter beside the host-owned ledger, and give the shared ledger filename helpers a lower-level, cycle-free ledger-format owner. Remove both back-edges without introducing replacement cycles, preserving ledger lifetime and recovery behaviour. This would take `host` out of the SCC; the existing `host -> node/enclave/...` consumer dependencies can remain.

**Then investigate `indexing`'s boundary.** Move [`HistoricalTransactionFetcher`](https://github.com/microsoft/CCF/blob/31f5f9fe14972cda34f7f44e258557f4dc574b51/src/indexing/historical_transaction_fetcher.h#L5-L24) beside its node-owned historical cache, and replace unnecessarily broad private KV includes in [`indexer.h`](https://github.com/microsoft/CCF/blob/31f5f9fe14972cda34f7f44e258557f4dc574b51/src/indexing/indexer.h#L5-L9), [`seqnos_by_key_bucketed.cpp`](https://github.com/microsoft/CCF/blob/31f5f9fe14972cda34f7f44e258557f4dc574b51/src/indexing/strategies/seqnos_by_key_bucketed.cpp#L4-L12), and [`visit_each_entry_in_map.cpp`](https://github.com/microsoft/CCF/blob/31f5f9fe14972cda34f7f44e258557f4dc574b51/src/indexing/strategies/visit_each_entry_in_map.cpp#L4-L6) with the narrow interfaces they actually use. **This is not an automatic SCC win:** direct includes of public `ccf/kv/...` headers still count as `kv` in this graph. A genuine acyclic boundary requires either a separately owned lower-level KV API or first removing `kv`'s own back-edges to `node` and `service`. Do not hide the dependency through transitive includes. Moving the historical fetcher alone leaves `indexing -> kv -> node -> indexing`.

No single-edge deletion shrinks the remaining 10-component SCC. The two highlighted pairs are the lowest-include-count **two-edge** cuts that do so in the measured graph, but include count is not implementation cost. Keep the broader configuration and KV/consensus work coordinated with #8309 and #8190; neither, nor #8117, is assumed merged here. Recompute the graph and enforce the resulting boundary after each real extraction.

### Measurement scope

Same source-folder projection as the preceding updates, using [`check-source-dependencies.py`](https://github.com/microsoft/CCF/blob/5e32dfbc4778018f5c2e377abdb125bc9186b808/scripts/check-source-dependencies.py) and its [policy/exclusion file](https://github.com/microsoft/CCF/blob/5e32dfbc4778018f5c2e377abdb125bc9186b808/scripts/source-dependencies.json), followed by strongly connected component analysis. Nodes are top-level `src` components with production C/C++ files. Public headers are resolved as include targets, but their bodies are not scanned as source vertices; public-only `ccf-api` and `research` targets, third-party/STL dependencies, tests, benchmarks, and fuzzers are omitted. This is not a complete transitive public-header or build-target graph, and the full DAG goal remains open.

**Comment by achamayou** [comment]
## 2026-09-17: host and enclave are acyclic; indexing is now a one-include cut

Measured on `main` at [`efc44e3`](https://github.com/microsoft/CCF/commit/efc44e31dacf013b362b4ce8c65f495f21489489) (#8396), using the same source-folder projection as the previous updates.

| State | Components | Direct edges | Components in cycles | Acyclic components | Outgoing policies |
| --- | ---: | ---: | ---: | ---: | ---: |
| 2026-09-13 projection, after #8338 | 25 | 104 | 10 | 15 | 15 |
| `main` at `efc44e3` | 25 | 96 | **8** | **17** | 17 |
| Proposed move 1: `indexing` | 25 | 95 | 7 | 18 | 18 |
| Proposed move 2: `common` (with or after #8309) | 25 | 92 | 6 | 19 | 19 |

The remaining **8-component SCC** is `common`, `consensus`, `endpoints`, `indexing`, `js`, `kv`, `node`, `service`.

The **17 acyclic components** are `apps`, `cose`, `crypto`, `ds`, `enclave`, `host`, `http`, `http_client`, `ledger`, `msgpack`, `pal`, `snapshots`, `tasks`, `tcp`, `threading`, `tls`, `uv`. All 17 have explicit outgoing policies in [`scripts/source-dependencies.json`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/scripts/source-dependencies.json). The full DAG goal remains open.

### Progress since the last update, per landed PR

| PR | Removed source-component edges | Added | Result |
| --- | --- | --- | --- |
| #8338 | `snapshots -> {consensus, host}` | none | 104 edges, 10 cyclic, exactly as projected |
| #7955 | `host -> indexing` | `indexing -> tasks` | LFS access now goes through `tasks` instead of the ringbuffer |
| #8117 | `enclave -> {tls, udp}`, `host -> udp`, `js -> tls`, `tcp -> ds`, and the old `tls`/`udp` leaf edges | `host -> {http, tasks}` | `udp` and the old TLS wrapper component are deleted: 23 components, 98 edges, still 10 cyclic |
| #8389 | `enclave -> host`, `node -> host`, `host -> {http, tasks, tcp}` | `host -> ledger`, `node -> ledger`, `enclave -> tls`, new `ledger` and `tls` leaf edges | **`host` leaves the core**; `ledger` and the new `tls` are born acyclic: 25 components, 100 edges, 9 cyclic |
| #8396 | `consensus -> enclave`, `node -> enclave`, `enclave -> {http, service, tcp, tls}` | `node -> {tcp, tls}` | **`enclave` leaves the core**: 96 edges, 8 cyclic |

A note on numbers: the descriptions of #8389 and #8396 report 26 components and 114 -> 110 edges because they also count the 14 include edges into the public-only `ccf-api` and `research` targets. This update keeps the issue's projection, in which those targets are resolved but not displayed: 96 edges (110 if they are included). The two views agree on the SCC.

### What changed structurally

With `enclave` gone, `node` is the **only** core component that still depends on `common`, `indexing`, or `js`. In the previous updates no single-edge deletion shrank the SCC; now there are three:

| Single-edge cut | Direct include witnesses | Frees | Note |
| --- | ---: | --- | --- |
| `node -> indexing` | 1 | `indexing` | [`node_state.h:34`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/src/node/node_state.h#L34) |
| `node -> common` | 4 | `common` | all four are `common/configuration.h`; overlaps #8309 |
| `node -> js` | 14 | `js` | governance handlers, `js_policy.h`, member/node frontends |

Each of these turns the freed component into an *outer consumer* of the core, exactly as #8389 and #8396 did for `host` and `enclave`. It does not make them lower-level "pure" components; that remains the goal for `kv` and `service`, see below.

Remaining core edges, with direct include witness counts (27 edges):

<details>
<summary>Core edge table</summary>

| Component | Outgoing core edges (witnesses) | Incoming core edges (witnesses) |
| --- | --- | --- |
| `common` | `consensus` (1), `node` (1), `service` (6) | `node` (4) |
| `consensus` | `kv` (5), `node` (6), `service` (5) | `common` (1), `node` (8), `service` (1) |
| `endpoints` | `kv` (1), `node` (7), `service` (12) | `js` (7), `node` (4) |
| `indexing` | `kv` (4), `node` (1) | `node` (1) |
| `js` | `endpoints` (7), `kv` (4), `node` (9), `service` (4) | `node` (14) |
| `kv` | `node` (4), `service` (5) | `consensus` (5), `endpoints` (1), `indexing` (4), `js` (4), `node` (24), `service` (7) |
| `node` | `common` (4), `consensus` (8), `endpoints` (4), `indexing` (1), `js` (14), `kv` (24), `service` (99) | `common` (1), `consensus` (6), `endpoints` (7), `indexing` (1), `js` (9), `kv` (4), `service` (3) |
| `service` | `consensus` (1), `kv` (7), `node` (3) | `common` (6), `consensus` (5), `endpoints` (12), `js` (4), `kv` (5), `node` (99) |

</details>

### Move 1 (now): take `indexing` out through `node -> indexing`

The single witness is [`node_state.h:34`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/src/node/node_state.h#L34) including `indexing/indexer.h`. `NodeState` uses the `Indexer` for exactly one thing: [`tick()`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/src/node/node_state.h#L2701-L2720) calls `indexer->update_strategies(elapsed, committed_txid)` while the node is `partOfNetwork`. The `Indexer` is constructed, installed as a subsystem, and owned by the composition root in [`enclave.h`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/src/enclave/enclave.h#L120-L123), whose [tick handler](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/src/enclave/enclave.h#L383-L385) already ticks the sibling `historical_state_cache` immediately after `node->tick(elapsed_ms)`.

1. **Drive `Indexer::update_strategies` from the composition root.** Remove the `indexer_` parameter from `NodeState::initialize` and the `indexer` member. In the enclave tick handler, directly after `node->tick(elapsed_ms)` (so still after `consensus->periodic`), call `indexer->update_strategies(elapsed_ms, committed)` gated on the existing `node->is_part_of_network()`. Expose the committed `TxID` from `NodeState` with a small accessor rather than reaching into consensus from `enclave.h`. Do **not** introduce a node-owned observer interface that `Indexer` implements: that would keep the pair coupled in the other direction and complicate a later lower-level `indexing`. The ordering relative to `n2n_channels->tick` changes, which does not affect the committed index.
2. **Enforce the boundary:** `"indexing": ["ccf-api", "crypto", "ds", "kv", "node", "tasks"]`, which is exactly the measured outgoing set. `indexing -> node` (the historical fetcher adapter) and `indexing -> kv` (transaction deserialisation in the strategies) remain as ordinary consumer dependencies. The 09-13 idea of making `indexing` lower than `kv` stays blocked on `kv`'s own back-edges and is not needed for this cut.
3. **Validation:** build the enclave/node targets and `indexing_test`; run `indexing_test` and the `e2e_logging.py` indexing paths (`test_committed_index`, `test_historical_query_range*`, `test_historical_query_sparse`), which exercise `SeqnosByKey_Bucketed` end to end. Recompute the graph and run the include/dependency, formatting and ASCII checks before pushing.

**Acceptance:** no production include from `node` into `indexing`; the recomputed graph has **95 edges, one 7-component SCC and 18 acyclic components**, all with policies. Projection, not a landed result.

### Move 2 (with or straight after #8309): take `common` out through `node -> common`

All four witnesses are `common/configuration.h`: [`history.h:11`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/src/node/history.h#L11) and [`rpc/frontend.h:16`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/src/node/rpc/frontend.h#L16) use nothing from it; [`node_call_types.h:80`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/src/node/rpc/node_call_types.h#L80) needs the JSON declarations for `StartupConfig::Start` (the `/node/create` body in `rpc/serialization.h`); [`node_interface.h:61`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/src/node/rpc/node_interface.h#L61) needs only the public `ccf/node/startup_config.h`. `node_state.h` also picks up `StartType` transitively through this header.

#8309 (draft, "Unify operator and node startup configuration") rewrites exactly this header: it introduces public `ccf/node/configuration.h` and `ccf/node/start_type.h`, but currently keeps the JSON declarations in `src/common/configuration.h` and keeps `node` including it, so it would leave `node -> common` in place. The cut is therefore best folded into #8309 or done immediately after it:

1. Place the `DECLARE_JSON_*` declarations for the public configuration types next to the types (public header, as `include/ccf/service/*` already does) or in a `node`-owned header, so that `node` never includes `common/configuration.h`; drop the two stale includes.
2. Take `StartType` out of `common/enclave_interface_types.h` into `ccf/node/start_type.h`, as #8309 already does.
3. Enforce `"common": ["ccf-api", "consensus", "crypto", "ds", "node", "pal", "service"]`. `common` then becomes what its consumers say it is: host/enclave shared configuration and ringbuffer types, consumed only by `host` and `enclave`.

**Acceptance:** with move 1, **92 edges, one 6-component SCC** (`consensus`, `endpoints`, `js`, `kv`, `node`, `service`), 19 acyclic components with policies.

### Prep that can land any time: relocate `service/internal_tables_access.h` to `node/`

[`internal_tables_access.h`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/src/service/internal_tables_access.h#L18-L22) is the only source of `service -> node` (3 witnesses: `node/history.h`, `node/ledger_secrets.h`, `node/uvm_endorsements.h`) and `service -> consensus` (1: `consensus/aft/raft_types.h`). Its eight includers are all under `src/node/`. Moving it removes both edges and adds none (96 -> 94 alone). It does not shrink the SCC by itself, because `service -> kv -> service` remains, but it leaves `service -> kv` as `service`'s only core out-edge, so `service` falls out automatically once `kv` does. Neither #8190 nor #8309 touches this file.

### After that: make `kv` pure, then `js`

`kv`'s only core out-edges are `kv -> node` (4: `node/rpc/claims.h` from `committable_tx.h` and `generic_serialise_wrapper.h`; `ccf/node/{cose_signatures_config,startup_config}.h` from `kv_types.h`) and `kv -> service` (5: `service/tables/{shares,signatures,snapshot_evidence}.h` from `deserialise.h`; `ccf/service/{consensus_type,reconfiguration_type}.h` from `kv_types.h`). Removing those nine witnesses, together with the prep above, would free `kv` **and** `service`, leaving a 4-component core (`consensus`, `endpoints`, `js`, `node`). This is the KV/consensus coupling the original issue asked to justify or remove, and it overlaps #8190 (`kv_types.h`, `store.h`), so it should be coordinated with that work rather than raced. `node -> js` (14 witnesses across governance, `js_policy.h` and the frontends) is the remaining single-edge cut, but it needs the constitution runtime behind an interface and is the most invasive of the three; it should wait until `kv` and `service` are settled.

<details>
<summary>Full refreshed source-component graph (25 components, 96 edges)</summary>

**Legend:** arrows run from an includer to its dependency. Dashed arrows participate in the remaining cycle; solid arrows do not. Green marks components that became acyclic since the last update (`host`, `enclave`, and the new `ledger` and `tls`). **Red nodes and red arrows mark the cuts proposed above**, not landed changes. Groupings are for readability, not dependency layers.

```mermaid
flowchart LR
  subgraph roots["Acyclic composition roots and consumers"]
    direction LR
    host["host/"]
    enclave["enclave/"]
    apps["apps/"]
  end
  subgraph core["Remaining cyclic core (8)"]
    direction LR
    common["common/"]
    consensus["consensus/"]
    endpoints["endpoints/"]
    indexing["indexing/"]
    js["js/"]
    kv["kv/"]
    node["node/"]
    service["service/"]
  end
  subgraph protocols["Acyclic protocols, transport and on-disk formats"]
    direction LR
    http["http/"]
    http_client["http_client/"]
    tls["tls/"]
    tcp["tcp/"]
    snapshots["snapshots/"]
    ledger["ledger/"]
  end
  subgraph support["Acyclic foundation and support"]
    direction LR
    crypto["crypto/"]
    ds["ds/"]
    pal["pal/"]
    tasks["tasks/"]
    cose["cose/"]
    msgpack["msgpack/"]
    threading["threading/"]
    uv["uv/"]
  end

  apps --> js
  common -.-> consensus
  common --> crypto
  common --> ds
  common -.-> node
  common --> pal
  common -.-> service
  consensus --> crypto
  consensus --> ds
  consensus -.-> kv
  consensus -.-> node
  consensus -.-> service
  crypto --> cose
  crypto --> ds
  enclave --> common
  enclave --> crypto
  enclave --> ds
  enclave --> indexing
  enclave --> js
  enclave --> kv
  enclave --> node
  enclave --> tasks
  endpoints --> crypto
  endpoints --> ds
  endpoints --> http
  endpoints -.-> kv
  endpoints -.-> node
  endpoints -.-> service
  host --> common
  host --> consensus
  host --> crypto
  host --> ds
  host --> enclave
  host --> http_client
  host --> kv
  host --> ledger
  host --> node
  host --> pal
  host --> service
  host --> snapshots
  host --> uv
  http --> crypto
  http --> ds
  http_client --> ds
  http_client --> uv
  indexing --> crypto
  indexing --> ds
  indexing -.-> kv
  indexing -.-> node
  indexing --> tasks
  js --> crypto
  js --> ds
  js -.-> endpoints
  js -.-> kv
  js -.-> node
  js --> pal
  js -.-> service
  kv --> crypto
  kv --> ds
  kv -.-> node
  kv -.-> service
  node -.-> common
  node -.-> consensus
  node --> cose
  node --> crypto
  node --> ds
  node -.-> endpoints
  node --> http
  node --> http_client
  node -.-> indexing
  node -.-> js
  node -.-> kv
  node --> ledger
  node --> pal
  node -.-> service
  node --> snapshots
  node --> tasks
  node --> tcp
  node --> tls
  pal --> crypto
  pal --> ds
  service -.-> consensus
  service --> cose
  service --> crypto
  service --> ds
  service -.-> kv
  service -.-> node
  service --> pal
  snapshots --> ds
  snapshots --> http
  snapshots --> http_client
  tasks --> ds
  tls --> crypto
  tls --> ds
  tls --> tasks
  tls --> tcp

  classDef opportunity fill:#ffebe9,stroke:#cf222e,stroke-width:3px,color:#82071e
  classDef extracted fill:#dafbe1,stroke:#1a7f37,stroke-width:2px,color:#116329
  class indexing,common opportunity
  class host,enclave,ledger,tls extracted
  linkStyle 61,69,81,86 stroke:#cf222e,stroke-width:3px
```

</details>

### Measurement scope

Same source-folder projection as the preceding updates: the scanner functions in [`scripts/check-source-dependencies.py`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/scripts/check-source-dependencies.py) and the exclusions and generated-header list in [`scripts/source-dependencies.json`](https://github.com/microsoft/CCF/blob/efc44e31dacf013b362b4ce8c65f495f21489489/scripts/source-dependencies.json), followed by strongly connected component analysis. Nodes are top-level `src` components containing production C/C++ sources or headers; tests, benchmarks and fuzzers are excluded. Public headers under `include/ccf` are resolved as include targets but not scanned as source vertices; the public-only `ccf-api` and `research` targets and third-party/STL dependencies are omitted from the displayed graph. Single- and two-edge cut candidates were enumerated over the 27 core edges; include witness counts are an ownership signal, not an implementation cost estimate. The checker enforces the configured allowlists through `scripts/includes-checks.sh`; it does not emit this graph or prove global acyclicity.

**Comment by achamayou** [comment]
## 2026-09-18: `indexing` is acyclic; the core is down to 7 components

Measured on `main` at [`f0ff3f7`](https://github.com/microsoft/CCF/commit/f0ff3f7be), using the same production source-folder projection as the previous updates (so the `ccf-api` policy pseudo-component is not counted as a source-folder edge).

| Metric | At `efc44e3` | Current |
| --- | ---: | ---: |
| Source components | 25 | 25 |
| Direct internal edges | 96 | 91 |
| Cyclic components | 8 | 7 |
| Acyclic components | 17 | 18 |
| Components with explicit policies | 17 | 18 |

#8398 has landed. It removed `node -> indexing` and moved `service/internal_tables_access.h` to `node/`, removing `service -> {consensus, node, cose, pal}`. `indexing` now has an explicit policy and is outside the remaining SCC. The production include graph has not changed since that merge through `f0ff3f7`.

The remaining cyclic core is:

`common`, `consensus`, `endpoints`, `js`, `kv`, `node`, `service`

### Suggested next cuts

1. **Finish `common` through #8309.** Remove the four remaining `node -> common` witnesses, all includes of `common/configuration.h`, by keeping public configuration JSON declarations with the public/node-owned types and removing the two stale includes. Then add the measured `common` policy. On today's graph, that direct cut gives **90 edges and a 6-component SCC**. Since #8309 rewrites these same headers, this should be folded into that PR or land immediately after it, not raced independently.
2. **Make `kv` a core library, coordinated with #8190.** Remove `kv -> node` and `kv -> service`, rather than trying to remove the natural consumer edge `service -> kv`:
   - move RPC claims and startup/config integration out of KV core;
   - move service-table-aware ledger deserialisation out of `kv/deserialise.h` into the node/service integration layer.

   With the `common` cut, removing those two direct edges frees both `kv` and `service`: projected **88 edges and a 4-component SCC** (`consensus`, `endpoints`, `js`, `node`). This is the highest-leverage next tranche and should add policies for both components.
3. **Put the constitution runtime behind a node-facing interface.** Removing `node -> js` then makes `js` acyclic: projected **87 edges and a 3-component SCC** (`consensus`, `endpoints`, `node`). At that point the remaining work is two explicit seams rather than a general tangle: `node <-> consensus` and `node <-> endpoints`.

These projections assume no unrelated source-folder edges are added; each landing PR should rerun the SCC analysis and add the newly valid policy in `scripts/source-dependencies.json`.
